contracts/VeJoeStaking.sol

```diff
@@ -119,12 +180,28 @@ contract VeJoeStaking is Initializable, OwnableUpgradeable {
         require(_amount > 0, "VeJoeStaking: expected deposit amount to be greater than zero");
 
         updateRewardVars();
-        // Transfer to the user their pending veJOE before updating their UserInfo
+
+        UserInfo storage userInfo = userInfos[msg.sender];
+
         if (_getUserHasNonZeroBalance(msg.sender)) {
+            // Transfer to the user their pending veJOE before updating their UserInfo
             _claim();
-        }
 
-        UserInfo storage userInfo = userInfos[msg.sender];
+            // We need to update user's `lastClaimTimestamp` to now to prevent
+            // passive veJOE accrual if user hit their max cap.
+            userInfo.lastClaimTimestamp = block.timestamp;
+
+            uint256 userStakedJoe = userInfo.balance;
+
+            // User is eligible for speed up benefits if `_amount` is at least
+            // `speedUpThreshold / 100 * userStakedJoe`
+            if (_amount.mul(100) >= speedUpThreshold.mul(userStakedJoe)) {
+                userInfo.speedUpEndTimestamp = block.timestamp.add(speedUpDuration);
+            }
+        } else {
+            userInfo.speedUpEndTimestamp = block.timestamp.add(speedUpDuration);
```

Comment:
A comment here could be useful saying if a user is making a deposit with no current balance they automatically get the speed up.